# This script is to generate plots about correlations between residuals and expresion levels of RBPDB RBPs
# This script uses RSEM gene RPKM data 
# 2023/06/21 made

# activate packages
library(ggplot2)
library(stringr)

# make directory which storage results generated by this script
setwd("C:/Rdata")
dir.create("20230621_CCLE_correlation_between_residual_and_RBPDB_RBP_RPKM")

# inport list about name of pri-miRNA, mature miRNA, and transcript
# this list is located at "//fsw-q02/okamura-lab/Files_related_to_M1_Projects/Hirota/CCLE_plot/not-depend_on_type_outlier_correlation_between_transcript_and_mature_miRNA"
setwd("C:/Rdata/not-depend_on_type_outlier_correlation_between_transcript_and_mature_miRNA")
name.df <-read.table("not-depend_on_type_outlier_summary_of_correlation_regarding_protein_coding.txt",sep="\t",header = T,stringsAsFactors = F)
name.df <-name.df[order(name.df[,1],name.df[,2],name.df[,3]),]

# set color palette
my.col <-colorRampPalette(c("orange","yellow","green","blue","red","brown","gray"))
my.col <-my.col(25)

# inport data about expression levels of RBPDB RBPs
# this data is located in "//fsw-q02/okamura-lab/Files_related_to_M1_Projects/Hirota/CCLE_Data"
setwd("C:/Rdata/CCLE_data")
RBP.exp <-read.table("RNAseq_of_RBP.txt",sep="\t",header = T,stringsAsFactors = F,check.names = F)
RBP <-colnames(RBP.exp)
RBP <-RBP[c(-1,-412)]
RBP <-sort(RBP)

# extract site primary
# data for extracting site primary is located at "//fsw-q02/okamura-lab/Files_related_to_M1_Projects/Hirota/CCLE_Data"
site <-sort(unique(RBP.exp[,412]))

# make table corresponded to site primary and clour
site_col <-as.data.frame(site)
site_col[,2] <-my.col
colnames(site_col) <-c("Site_Primary","color")

# list files of residulas
# these files are located at ""
r.file <-list.files(path = "C:/Rdata/result_of_calculating_residual",pattern = ".txt")
r.file <-r.file[r.file!="hsa-mir-640_hsa-miR-640_vs_ATAD2.txt"]

# make empty table for writting summary
sm <-as.data.frame(matrix(nrow=410*289,ncol=7))
colnames(sm) <-c("pri-miRNA","miRNA","primary_transcript","RBP","R","p.value","number_of_cell")

#generate plots and summary about correlations between residuals and expression levels of RBPDB RBPs
for (i in 1:length(r.file)) {
  for (k in 1:length(RBP)) {
    if(k==1) {
      # make and name folder
      setwd("C:/Rdata/20230621_CCLE_correlation_between_residual_and_RBPDB_RBP_RPKM")
      f.name <-paste0(name.df[i,1],"_",name.df[i,2],"_",name.df[i,3])
      dir.create(f.name)
    }
    
    # read a file of residual 
    setwd("C:/Rdata/result_of_calculating_residual")
    res.df <-read.table(r.file[i],sep="\t",header = T,stringsAsFactors = F,check.names = F)
    cell <-res.df[,4]
    res.df <-res.df[,c(-1,-2)]
    
    # extract rows of cell that exsit residual data from RBP expression data 
    m.row <-match(RBP[k],colnames(RBP.exp))
    RBP.df <-RBP.exp[,c(1,m.row,412)]
    RBP.df <-RBP.df[,c(2,1,3)]
    
    #merge residual data and RBP expression data
    merged.df <-merge(res.df,RBP.df,by=c("CELL","Site_Primary"))
    merged.df <-merged.df[,c(3,4,1,2)]
    
    # normalize RBP expression by log2
    merged.df[,2] <-merged.df[,2]+1
    merged.df[,2] <-log2(merged.df[,2])
    
    # merge with color data frame 
    merged.df <-merge(merged.df,site_col,by="Site_Primary")
    merged.df <-merged.df[,c(2:4,1,5)]
    
    # investigate correlation coefficient
    r <-try(cor.test(merged.df[,1],merged.df[,2],method = "spearman"),silent = T)
    
    if(class(r)!="try-error"){
      
      #generate plot 
      setwd(paste0("C:/Rdata/20230621_CCLE_correlation_between_residual_and_RBPDB_RBP_RPKM/",f.name))
      cp <-ggplot(data=merged.df,aes(x=merged.df[,1],y=merged.df[,2]))+
        geom_point(aes(color=Site_Primary))+ #draw scatter plot
        stat_smooth(method = "lm", se = FALSE, colour = "black", linewidth = 0.5)+ # draw regression line
        scale_color_manual(values =unique(merged.df[,5]))+ # color by site primary
        labs(title=paste0("R =",signif(r$estimate,3),", p = ",signif(r$p.value,3),", n = ",nrow(merged.df)),x=paste0("biogenesis efficiency of ",name.df[i,1],"_",name.df[i,2],"_",name.df[i,3]),
             y=paste0("Log2 (RPKM+1) of ",RBP[k]))+ #write main title and la title
        guides(color = guide_legend(ncol = 1))+ #make legend 1line
        theme(legend.text = element_text(size=5)) + #change size 
        theme_bw()+
        theme(legend.background = element_rect(fill = "white", colour = "black"))
      
      cp <-cp+
        theme(plot.title=element_text(hjust = 0.5),axis.title = element_text(size = 15),axis.text = element_text(size = 13),
              title = element_text(size = 15),legend.title = element_text(hjust = 0.5))
      
      ggsave(paste0(name.df[i,1],"_",name.df[i,2],"_",name.df[i,3],"_vs_",RBP[k],".pdf"),width =10 ,height =7.5 )
      
      # write summary satisfy condition 
      sm[410*(i-1)+k,1] <-name.df[i,1]
      sm[410*(i-1)+k,2] <-name.df[i,2]
      sm[410*(i-1)+k,3] <-name.df[i,3]
      sm[410*(i-1)+k,4] <-RBP[k]
      sm[410*(i-1)+k,5] <-r$estimate
      sm[410*(i-1)+k,6] <-r$p.value
      sm[410*(i-1)+k,7] <-nrow(merged.df)
      
    }else{
      # write summary don't satisfy condition 
      sm[410*(i-1)+k,1] <-name.df[i,1]
      sm[410*(i-1)+k,2] <-name.df[i,2]
      sm[410*(i-1)+k,3] <-name.df[i,3]
      sm[410*(i-1)+k,4] <-RBP[k]
      sm[410*(i-1)+k,5] <-NA
      sm[410*(i-1)+k,6] <-NA
      sm[410*(i-1)+k,7] <-nrow(merged.df)
    }
    
    
    if(i==289&&k==410){
      #export summary at the end of this script
      sm <-sm[order(sm[,5],sm[,7],decreasing = T),]
      setwd("C:/Rdata/20230621_CCLE_correlation_between_residual_and_RBPDB_RBP_RPKM")
      write.table(sm,"summary_of_correlation_between_residual_and_RBPDB_RBP_RPKM.txt",sep="\t",row.names = F,quote = F)
    }
  }}